Allow use of multiple particle load balancing techniques

gassmoeller opened the pull request:
In hindsight it was wrong to assume that a user would only want to use one particle load balancing strategy. If for example one has a wide spectrum of allowed number of particles per cell, one might still want to repartition the mesh according to the combined weights of cells and particles (this is what I want to do right now in a model).  This PR allows to combine the existing strategies. It also allows to only add particles in poorly populated areas, which was previously only possible if the remove particles strategy was selected as well. Finally, I would like to change the default balancing strategy to 'repartition', since it does not harm a model (it is a small slowdown of the repartition, but it can lead to significant improvements in runtime), and it speeds up models for users with parallel models who want to 'just switch on the particles' without thinking about the details.

One thing I noticed in the tests is that apparently the partition influences the mesh refinement, e.g. I get different meshes when they are partitioned differently. Is that something to worry about, or is it expected? And should I therefore leave the change of the default value for another PR?

bangerth commented on source/particle/world.cc:
```diff
@@ -1279,10 +1281,14 @@ namespace aspect
+                             "integrator and particle properties. In general for a more "
+                             "expensive integrator and more expensive properties a larger "
+                             "particle weight is recommended. Every cell without particles "
+                             "is associated with a weight of 1000.");
```
Maybe rephrase this as "Before adding the weights of particles, each cell already carries a weight of 1000 to account for the cost of field-based computations."

bangerth commented on source/particle/world.cc:
```diff
@@ -1328,16 +1334,18 @@ namespace aspect
+
+          if (std::find(strategies.begin(),strategies.end(), "remove particles") != strategies.end())
+            particle_load_balancing = typename ParticleLoadBalancing::Kind(particle_load_balancing | ParticleLoadBalancing::remove_particles);
+          if (std::find(strategies.begin(),strategies.end(), "add particles") != strategies.end())
```
this blocks is dangerous since it does not ensure that we take care of all elements of the list. I think it would be better to write it as a loop over all elements of the list, and for each one test it against all known possibilities via an `if ... else if ...` sequence that ends with an `AssertThrow (false, ...` for the unrecognized element of the list.

bangerth commented on tests/particle_timing_information/screen-output:
```diff
@@ -27,28 +27,28 @@ Number of degrees of freedom: 447 (208+31+104+104)
 
-Number of active cells: 16 (on 4 levels)
-Number of degrees of freedom: 387 (180+27+90+90)
+Number of active cells: 31 (on 4 levels)
```
what caused the change in the mesh?

gassmoeller commented:
I addressed the comments. 
The change in the test is caused by a different mesh partition (because the test uses the default value, which changed from 'none' to 'repartition'). A few of the cells that are now on a different process got refined. When I change the parameter file to use the old partition strategy the test remains unchanged (would you prefer that over the current change? We have another separate test for the mesh partition).

bangerth commented:
OK, good enough. I don't care either way about changing the test or forcing it to be the same as before, as long as you feel that all code paths that were tested before are still tested one way or the other.

bangerth commented on source/particle/world.cc:
```diff
@@ -1328,16 +1335,33 @@ namespace aspect
+                particle_load_balancing = typename ParticleLoadBalancing::Kind(particle_load_balancing | ParticleLoadBalancing::repartition);
+              else if (*strategy == "none")
+                {
+                  particle_load_balancing = ParticleLoadBalancing::no_balancing;;
```
oh, while not wrong, the second semicolon is still unnecessary. please remove before merging.

bangerth commented on source/particle/world.cc:
```diff
@@ -1328,16 +1335,33 @@ namespace aspect
+              else
+                AssertThrow(false,
+                            ExcMessage("The <Load balancing strategy> parameter contains an unknown value that "
+                                       "does not correspond to any know load balancing strategy."));
```
can you print the offending value's string?

gassmoeller commented:
PR is rebased and tests pass again. Should be ready to merge.

bangerth commented on source/particle/world.cc:
```diff
@@ -1317,9 +1319,9 @@ namespace aspect
-                                                  "remove and add particles|repartition"),
+          prm.declare_entry ("Load balancing strategy", "repartition",
+                             Patterns::MultipleSelection ("none|remove particles|add particles|"
+                                                          "remove and add particles|repartition"),
```
In a follow-up patch, would you mind documenting here what the different choices imply? Also, what it means to list more than one of them.
